### PR TITLE
Potential changes to support no IOperationSource<T> from an LRO which inherits Operation

### PR DESCRIPTION
**Pull request by m-nash**
# Description

- Added OperationInternals without the T
- Added OperationOrResponseInternals without the T
- Update lro test cases since the non T versions no longer have WaitForCompletionAsync they instead of WaitForCompletionResponseAsync
- Update LROWriter to write the appropriate T or non T based on Result type
- Removed the implementation of IOperationSource<T> when the implementation is non T since there is no result to create

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs [hunk 1]
```diff
@@ -70,13 +70,21 @@ void WriteResultFunction(bool async)
             var @namespace = cs.Namespace;
             using (writer.Namespace(@namespace))
             {
+                bool isGenericOperation = !(operation.ResultType.IsFrameworkType && operation.ResultType.FrameworkType == typeof(Response));
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs:73]: We should fix the logic in https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Common/Output/Models/LongRunningOperation.cs#L52 instead.

File: src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs [hunk 2]
```diff
@@ -103,7 +111,7 @@ void WriteResultFunction(bool async)
 
                     using (writer.Scope())
                     {
-                        writer.Line($"_operation = new {helperType}(this, clientDiagnostics, pipeline, request, response, {typeof(OperationFinalStateVia)}.{operation.FinalStateVia}, {operation.Diagnostics.ScopeName:L});");
+                        writer.Line($"_operation = new {helperType}{paramStart}clientDiagnostics, pipeline, request, response, {typeof(OperationFinalStateVia)}.{operation.FinalStateVia}, {operation.Diagnostics.ScopeName:L});");
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs:114]: You can split the `Line` call into multiple `Append` and have the `                var paramStart = isGenericOperation ? "(this, " : "(";` logic inline.

```C#
if (isGenericOperation)
{
writer.Append("this,");
}

File: src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs [hunk 3]
```diff
@@ -70,13 +70,21 @@ void WriteResultFunction(bool async)
...
+                var waitForCompletionType = new CSharpType(typeof(ValueTask<>), valueTaskType);
+                var helperType = isGenericOperation ? new CSharpType(typeof(OperationInternals<>), operation.ResultType) : new CSharpType(typeof(OperationInternals));
+                var paramStart = isGenericOperation ? "(this, " : "(";
+                var waitForCompleteMethodName = isGenericOperation ? "WaitForCompletionAsync" : "WaitForCompletionResponseAsync";
+
+                FormattableString scope = isGenericOperation
+                    ? (FormattableString)$"{operation.Declaration.Accessibility} partial class {cs.Name}: {baseType}, {interfaceType}"
+                    : (FormattableString)$"{operation.Declaration.Accessibility} partial class {cs.Name}: {baseType}";
+
+                using (writer.Scope(scope))
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs:87]: You can move the `$"{operation.Declaration.Accessibility} partial class {cs.Name}: {baseType}` before the scope call and have the logic inline.

``` C#
writer.Append("$"{operation.Declaration.Accessibility} partial class {cs.Name}: {baseType});
if (isGenericOperation)
{
	writer.Append(", interfaceType");
}
using (writer.Scope()) {}
```

Comment (pakrym) [reply src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs:87]: Basic idea is to have as little non-linearity in writing as possible.

File: src/assets/Generator.Shared/OperationInternals.cs [hunk 4]
```diff
@@ -127,14 +121,11 @@ private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToke
...
-                    {
-                        _value = async
-                            ? await _source.CreateResultAsync(finalResponse, cancellationToken).ConfigureAwait(false)
-                            : _source.CreateResult(finalResponse, cancellationToken);
-                        _rawResponse = finalResponse;
-                        _hasValue = true;
-                        break;
-                    }
+                        {
+                            await SetValueAsync(async,  finalResponse, cancellationToken);
```
Comment (pakrym) [inline src/assets/Generator.Shared/OperationInternals.cs:125]:
```suggestion
                            await SetValueAsync(async, finalResponse, cancellationToken);
```

File: src/assets/Generator.Shared/OperationOrResponseInternals.cs [hunk 5]
```diff
@@ -19,70 +19,22 @@ namespace Azure.Core
     /// https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation
     /// https://github.com/Azure/adx-documentation-pr/blob/master/sdks/LRO/LRO_AzureSDK.md
     /// </summary>
-    /// <typeparam name="T">The final result of the LRO.</typeparam>
-    internal class OperationOrResponseInternals<T>
+    internal class OperationOrResponseInternals : OperationOrResponseInternalsBase
```
Comment (pakrym) [inline src/assets/Generator.Shared/OperationOrResponseInternals.cs:22]: Can we merge this type and the OperationOrResponseInternalsBase?

**Review by pakrym — APPROVED** [review]
Very nice! Thank you for changing the data-plane code-gen too!

**Review by allenjzhang — APPROVED** [review]
:shipit:

**Comment by ShivangiReja** [comment]
@m-nash can we move LongRunningOperationWriter.cs file inside the [Writers folder](https://github.com/Azure/autorest.csharp/tree/feature/v3/src/AutoRest.CSharp/Common/Generation) ?

File: src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs [hunk 6]
```diff
@@ -70,21 +70,21 @@ void WriteResultFunction(bool async)
...
-                var paramStart = isGenericOperation ? "(this, " : "(";
-                var waitForCompleteMethodName = isGenericOperation ? "WaitForCompletionAsync" : "WaitForCompletionResponseAsync";
+                var helperType = operation.ResultType != null ? new CSharpType(typeof(OperationInternals<>), operation.ResultType) : new CSharpType(typeof(OperationInternals));
+                var waitForCompleteMethodName = operation.ResultType != null ? "WaitForCompletionAsync" : "WaitForCompletionResponseAsync";
 
-                FormattableString scope = isGenericOperation
-                    ? (FormattableString)$"{operation.Declaration.Accessibility} partial class {cs.Name}: {baseType}, {interfaceType}"
-                    : (FormattableString)$"{operation.Declaration.Accessibility} partial class {cs.Name}: {baseType}";
+                writer.Append($"{operation.Declaration.Accessibility} partial class {cs.Name}: {baseType}");
+                if (operation.ResultType != null)
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Generation/LongRunningOperationWriter.cs:82]:
```suggestion
                if (interfaceType != null)
```

**Pull request merged**